proposals/oci-support-updates.md

```diff
@@ -0,0 +1,95 @@
+# Proposal: OCI Support Updates
+
+A couple of the maintainers recently met regarding how to move forward with Helm's experimental OCI support. This documents the conclusions reached.
+
+Note: the existing support for OCI registries is documented [here](https://helm.sh/docs/topics/registries/).
+
+## 1. Implement `Getter` and introduce `Pusher`
+
+As of yet, the code related to registry support has been written standalone (see `internal/experimental/registry`).
+
+The act of downloading a chart from an OCI registry should mimic what is already possible using downloader plugins. In theory, using a protocol prefix such as `oci://` in any place where chart repos are referenced should work just by implementing a `Getter` for OCI.
+
+Currently there is no Helm-specific way to upload chart packages. Using the same model of `Getter`, a new interface called `Pusher` should be introduced, with a single built-in implementation (OCI). This opens the door for plugins to implement their own upload mechanism, and for Helm to add a new top-level upload command such as `helm push`.
+
+
+## 2. Support for provenance files
+
+Helm currently has the ability to verify package signatures, assuming the presence of a file suffixed with `.prov` sitting next to a chart `.tgz` in a repository.
+
+Support for this method of signature validation should be carried over into OCI storage. As the format of the OCI manifest is custom to Helm, Helm can also choose to modify the resulting manifest on upload when a chart is being signed (e.g. `helm push --sign`).
+
+As far as the low-level details, the `.prov` file will simply be stored as a second layer on the manifest. If running `helm pull --verify oci://...`, the second layer (layer at index 1) will be assumed to be the provenance file. The first layer is always the chart `.tgz` itself.
+
+Here is an example of what a manifest might look like with a provenance file attached:
+
+```
+{
+  "schemaVersion": 2,
+  "config": {
+    "mediaType": "application/vnd.cncf.helm.config.v1+json",
+    "digest": "sha256:8ec7c0f2f6860037c19b54c3cfbab48d9b4b21b485a93d87b64690fdb68c2111",
+    "size": 117
+  },
+  "layers": [
+    {
+      "mediaType": "application/tar+gzip",
```

Comment:
As per https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#defining-layermediatypes, media types that are not `vendor|org|entity` specific should be registered with IANA.

I don't see `application/tar+gzip` in this list; is there an existing generic media type that would be appropriate to use here, or should we come up with a new non-generic type? https://www.iana.org/assignments/media-types/media-types.xhtml

Reply:
The `+gzip` suffix is [registered](https://www.iana.org/assignments/media-type-structured-suffix/media-type-structured-suffix.xhtml), but there [doesn't appear to be a registered media type for tar](https://stackoverflow.com/questions/39058006/proper-mime-media-type-for-tar-files).

Does it make sense to use a Helm vendor-defined type for the layer, since that means we can assume more about the archive than 'tarball', e.g. presence of _Chart.yaml_.

Reply:
Hm. This is an unfortunate mistake. cc @stevelasker

I'm also reading this: https://superuser.com/a/960710

I'm not sure I agree it should be a custom type, but this may be what we have to do if tar is really not an official type..

Reply:
For they layer types, this was another active conversation. Should each type define their own `layer.mediaType`. When it's as simple as a `.tar` file, there's not a ton of benefit for defining something unique. As noted, `application/tar`, nor `application/tar+gzip` have yet to be registered with iana.org. It was a surprise to all of us.

The reality is the layer is less of a concern. Artifact types must be unique at the `manifest.config` and hopefully soon `index.config` layer as tooling must know if it can operate on the artifact. As a security scanner iterates through the artifacts in a registry, if and how will they scan the artifact?
The layer is a secondary object that's determined by the specific artifact tool.

If someone wants to register `application/tar` with iana.org, it would be much appreciated, and you'll get your name in history - a reason to do, or not to do it...